float8_experimental/fsdp_utils.py

```diff
@@ -0,0 +1,61 @@
+import math
+import warnings
+from typing import List
+
+import torch
+import torch.nn as nn
+from float8_experimental.float8_dynamic_utils import WeightWithDynamicFloat8CastTensor
+from float8_experimental.float8_linear import Float8Linear
+from float8_experimental.float8_linear_utils import linear_requires_sync
+from float8_experimental.float8_utils import EPS
+
+
+def precompute_float8_scale_for_fsdp(module: nn.Module) -> None:
+    """
+    Calculate scale for all float8 parameters after optimizer step
+    It performs a single all-reduce instead of many all-reduces for each parameter
+    Exmaple usage:
+        model(input).sum().backward()
+        optim.step()
+        precompute_float8_scale_for_fsdp(model)
+    """
+    from torch.distributed._tensor import DTensor
+
+    if any(
+        isinstance(m, Float8Linear)
+        and linear_requires_sync(
+            m.scaling_type_x, m.scaling_type_w, m.scaling_type_dL_dY
+        )
+        for m in module.modules()
+    ):
+        raise NotImplementedError("Only supports delayed scaling")
+    float8_linears: List[Float8Linear] = [
```

Comment:
is this expensive for real models? if yes, maybe we can offer option to precompute this?

Reply:
My intuition is that this should be pretty fast as the number of `nn.Module`s in the model is usually at most in the thousands and this is pure Python overhead. @weifengpy you can check the traces you have if you see any noticeable gaps from this.

Reply:
just checked the profiler traces. it's roughly 0.15ms cpu overhead (5% of `precompute_float8_dynamic_scale_for_fsdp` and is tiny portion of 1 training loop). no cuda are launched

thus I am keeping it as is now for simplicity

![Uploading Screenshot 2024-07-11 at 2.45.17 PM.png…]()